// million@1.12.3/dist/html.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const utils = require('./chunks/render.cjs');

const factory = (tagName) => {
  function vnode(param1, param2) {
    if (Array.isArray(param1)) {
      return utils.h(tagName, param2, ...param1);
    }
    return utils.h(tagName, param1, ...param2 ?? []);
  }
  return vnode;
};
const input = factory("input");
const textarea = factory("textarea");
const checkbox = factory("input");
const address = factory("address");
const article = factory("article");
const aside = factory("aside");
const footer = factory("footer");
const header = factory("header");
const h1 = factory("h1");
const h2 = factory("h2");
const h3 = factory("h3");
const h4 = factory("h4");
const h5 = factory("h5");
const h6 = factory("h6");
const hgroup = factory("hgroup");
const nav = factory("nav");
const section = factory("section");
const blockquote = factory("blockquote");
const dd = factory("dd");
const div = factory("div");
const dl = factory("dl");
const dt = factory("dt");
const figcaption = factory("figcaption");
const figure = factory("figure");
const hr = factory("hr");
const li = factory("li");
const main = factory("main");
const ol = factory("ol");
const p = factory("p");
const pre = factory("pre");
const ul = factory("ul");
const a = factory("a");
const abbr = factory("abbr");
const b = factory("b");
const bdi = factory("bdi");
const bdo = factory("bdo");
const br = factory("br");
const cite = factory("cite");
const code = factory("code");
const data = factory("data");
const dfn = factory("dfn");
const em = factory("em");
const i = factory("i");
const kbd = factory("kbd");
const mark = factory("mark");
const q = factory("q");
const rp = factory("rp");
const rt = factory("rt");
const rtc = factory("rtc");
const ruby = factory("ruby");
const s = factory("s");
const samp = factory("samp");
const small = factory("small");
const span = factory("span");
const strong = factory("strong");
const sub = factory("sub");
const sup = factory("sup");
const time = factory("time");
const u = factory("u");
const varElement = factory("var");
const wbr = factory("wbr");
const area = factory("area");
const audio = factory("audio");
const img = factory("img");
const map = factory("map");
const track = factory("track");
const video = factory("video");
const embed = factory("embed");
const object = factory("object");
const param = factory("param");
const picture = factory("picture");
const source = factory("source");
const canvas = factory("canvas");
const script = factory("script");
const del = factory("del");
const ins = factory("ins");
const caption = factory("caption");
const col = factory("col");
const colgroup = factory("colgroup");
const table = factory("table");
const tbody = factory("tbody");
const td = factory("td");
const tfoot = factory("tfoot");
const th = factory("th");
const thead = factory("thead");
const tr = factory("tr");
const button = factory("button");
const datalist = factory("datalist");
const fieldset = factory("fieldset");
const form = factory("form");
const label = factory("label");
const legend = factory("legend");
const meter = factory("meter");
const optgroup = factory("optgroup");
const option = factory("option");
const output = factory("output");
const progress = factory("progress");
const select = factory("select");
const details = factory("details");
const menuitem = factory("menuitem");
const summary = factory("summary");
const slot = factory("slot");
const template = factory("template");
const circle = factory("circle");
const rect = factory("rect");
const ellipse = factory("ellipse");
const g = factory("g");
const image = factory("image");
const line = factory("line");
const mask = factory("mask");
const path = factory("path");
const polygon = factory("polygon");
const polyline = factory("polyline");
const svg = factory("svg");
const svgText = factory("text");
const marker = factory("marker");
const linearGradient = factory("linearGradient");
const foreignObject = factory("foreignObject");

exports.build = utils.build;
exports.evaluate = utils.evaluate;
exports.htm = utils.htm;
exports.html = utils.html;
exports.treeify = utils.treeify;
exports.a = a;
exports.abbr = abbr;
exports.address = address;
exports.area = area;
exports.article = article;
exports.aside = aside;
exports.audio = audio;
exports.b = b;
exports.bdi = bdi;
exports.bdo = bdo;
exports.blockquote = blockquote;
exports.br = br;
exports.button = button;
exports.canvas = canvas;
exports.caption = caption;
exports.checkbox = checkbox;
exports.circle = circle;
exports.cite = cite;
exports.code = code;
exports.col = col;
exports.colgroup = colgroup;
exports.data = data;
exports.datalist = datalist;
exports.dd = dd;
exports.del = del;
exports.details = details;
exports.dfn = dfn;
exports.div = div;
exports.dl = dl;
exports.dt = dt;
exports.ellipse = ellipse;
exports.em = em;
exports.embed = embed;
exports.factory = factory;
exports.fieldset = fieldset;
exports.figcaption = figcaption;
exports.figure = figure;
exports.footer = footer;
exports.foreignObject = foreignObject;
exports.form = form;
exports.g = g;
exports.h1 = h1;
exports.h2 = h2;
exports.h3 = h3;
exports.h4 = h4;
exports.h5 = h5;
exports.h6 = h6;
exports.header = header;
exports.hgroup = hgroup;
exports.hr = hr;
exports.i = i;
exports.image = image;
exports.img = img;
exports.input = input;
exports.ins = ins;
exports.kbd = kbd;
exports.label = label;
exports.legend = legend;
exports.li = li;
exports.line = line;
exports.linearGradient = linearGradient;
exports.main = main;
exports.map = map;
exports.mark = mark;
exports.marker = marker;
exports.mask = mask;
exports.menuitem = menuitem;
exports.meter = meter;
exports.nav = nav;
exports.object = object;
exports.ol = ol;
exports.optgroup = optgroup;
exports.option = option;
exports.output = output;
exports.p = p;
exports.param = param;
exports.path = path;
exports.picture = picture;
exports.polygon = polygon;
exports.polyline = polyline;
exports.pre = pre;
exports.progress = progress;
exports.q = q;
exports.rect = rect;
exports.rp = rp;
exports.rt = rt;
exports.rtc = rtc;
exports.ruby = ruby;
exports.s = s;
exports.samp = samp;
exports.script = script;
exports.section = section;
exports.select = select;
exports.slot = slot;
exports.small = small;
exports.source = source;
exports.span = span;
exports.strong = strong;
exports.sub = sub;
exports.summary = summary;
exports.sup = sup;
exports.svg = svg;
exports.svgText = svgText;
exports.table = table;
exports.tbody = tbody;
exports.td = td;
exports.template = template;
exports.textarea = textarea;
exports.tfoot = tfoot;
exports.th = th;
exports.thead = thead;
exports.time = time;
exports.tr = tr;
exports.track = track;
exports.u = u;
exports.ul = ul;
exports.varElement = varElement;
exports.video = video;
exports.wbr = wbr;
