// million@1.12.3/dist/html.mjs
import { x as h } from './chunks/render.mjs';
export { K as build, J as evaluate, G as htm, H as html, I as treeify } from './chunks/render.mjs';

const factory = (tagName) => {
  function vnode(param1, param2) {
    if (Array.isArray(param1)) {
      return h(tagName, param2, ...param1);
    }
    return h(tagName, param1, ...param2 ?? []);
  }
  return vnode;
};
const input = factory("input");
const textarea = factory("textarea");
const checkbox = factory("input");
const address = factory("address");
const article = factory("article");
const aside = factory("aside");
const footer = factory("footer");
const header = factory("header");
const h1 = factory("h1");
const h2 = factory("h2");
const h3 = factory("h3");
const h4 = factory("h4");
const h5 = factory("h5");
const h6 = factory("h6");
const hgroup = factory("hgroup");
const nav = factory("nav");
const section = factory("section");
const blockquote = factory("blockquote");
const dd = factory("dd");
const div = factory("div");
const dl = factory("dl");
const dt = factory("dt");
const figcaption = factory("figcaption");
const figure = factory("figure");
const hr = factory("hr");
const li = factory("li");
const main = factory("main");
const ol = factory("ol");
const p = factory("p");
const pre = factory("pre");
const ul = factory("ul");
const a = factory("a");
const abbr = factory("abbr");
const b = factory("b");
const bdi = factory("bdi");
const bdo = factory("bdo");
const br = factory("br");
const cite = factory("cite");
const code = factory("code");
const data = factory("data");
const dfn = factory("dfn");
const em = factory("em");
const i = factory("i");
const kbd = factory("kbd");
const mark = factory("mark");
const q = factory("q");
const rp = factory("rp");
const rt = factory("rt");
const rtc = factory("rtc");
const ruby = factory("ruby");
const s = factory("s");
const samp = factory("samp");
const small = factory("small");
const span = factory("span");
const strong = factory("strong");
const sub = factory("sub");
const sup = factory("sup");
const time = factory("time");
const u = factory("u");
const varElement = factory("var");
const wbr = factory("wbr");
const area = factory("area");
const audio = factory("audio");
const img = factory("img");
const map = factory("map");
const track = factory("track");
const video = factory("video");
const embed = factory("embed");
const object = factory("object");
const param = factory("param");
const picture = factory("picture");
const source = factory("source");
const canvas = factory("canvas");
const script = factory("script");
const del = factory("del");
const ins = factory("ins");
const caption = factory("caption");
const col = factory("col");
const colgroup = factory("colgroup");
const table = factory("table");
const tbody = factory("tbody");
const td = factory("td");
const tfoot = factory("tfoot");
const th = factory("th");
const thead = factory("thead");
const tr = factory("tr");
const button = factory("button");
const datalist = factory("datalist");
const fieldset = factory("fieldset");
const form = factory("form");
const label = factory("label");
const legend = factory("legend");
const meter = factory("meter");
const optgroup = factory("optgroup");
const option = factory("option");
const output = factory("output");
const progress = factory("progress");
const select = factory("select");
const details = factory("details");
const menuitem = factory("menuitem");
const summary = factory("summary");
const slot = factory("slot");
const template = factory("template");
const circle = factory("circle");
const rect = factory("rect");
const ellipse = factory("ellipse");
const g = factory("g");
const image = factory("image");
const line = factory("line");
const mask = factory("mask");
const path = factory("path");
const polygon = factory("polygon");
const polyline = factory("polyline");
const svg = factory("svg");
const svgText = factory("text");
const marker = factory("marker");
const linearGradient = factory("linearGradient");
const foreignObject = factory("foreignObject");

export { a, abbr, address, area, article, aside, audio, b, bdi, bdo, blockquote, br, button, canvas, caption, checkbox, circle, cite, code, col, colgroup, data, datalist, dd, del, details, dfn, div, dl, dt, ellipse, em, embed, factory, fieldset, figcaption, figure, footer, foreignObject, form, g, h1, h2, h3, h4, h5, h6, header, hgroup, hr, i, image, img, input, ins, kbd, label, legend, li, line, linearGradient, main, map, mark, marker, mask, menuitem, meter, nav, object, ol, optgroup, option, output, p, param, path, picture, polygon, polyline, pre, progress, q, rect, rp, rt, rtc, ruby, s, samp, script, section, select, slot, small, source, span, strong, sub, summary, sup, svg, svgText, table, tbody, td, template, textarea, tfoot, th, thead, time, tr, track, u, ul, varElement, video, wbr };
