// million@1.12.3/dist/million.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const utils = require('./chunks/render.cjs');

const _ = void 0;

exports.DOM_REF_FIELD = utils.DOM_REF_FIELD;
exports.DeltaTypes = utils.DeltaTypes;
exports.Deltas = utils.Deltas;
exports.EffectTypes = utils.EffectTypes;
exports.Flags = utils.Flags;
exports.OLD_VNODE_FIELD = utils.OLD_VNODE_FIELD;
exports.batch = utils.batch;
exports.className = utils.className;
exports.createElement = utils.createElement;
exports.diff = utils.diff;
exports.flushQueue = utils.flushQueue;
exports.hydrate = utils.hydrate;
exports.isPending = utils.isPending;
exports.kebab = utils.kebab;
exports.m = utils.m;
exports.mergeHooks = utils.mergeHooks;
exports.ns = utils.ns;
exports.patch = utils.patch;
exports.render = utils.render;
exports.startTransition = utils.startTransition;
exports.style = utils.style;
exports.svg = utils.svg;
exports.updateProp = utils.updateProp;
exports.useChildren = utils.useChildren;
exports.useNode = utils.useNode;
exports.useProps = utils.useProps;
exports._ = _;
